### PR TITLE
Add `testQuick` command

### PR DESCRIPTION
**Pull request by HollandDM**
Add `testQuick` command that leverage `CallGraphAnalysis`.

This pull request introduces the `testQuick` command, which utilizes `CallGraphAnalysis`. The following changes have been implemented:
- The outputs of `CallGraphAnalysis` now include `invalidClassNames`, which contains the names of test classes extracted from the `spanningInvalidationTree`. This list indicates the classes that are affected by code changes.
- A `callGraphAnalysis` task has been added to `JavaModule`. This is a persistent task that, when executed, generates a `callGraphAnalysis` folder containing data related to the `CallGraphAnalysis` feature.
- The `codeSignatures` of `MillBuildRootModule` have been updated to incorporate the results from `callGraphAnalysis`.
- The `testQuick` command has been added to `JavaTests`. It uses the `invalidClassNames` output from `callGraphAnalysis` and a dedicated log file, `quickTestFailedClasses.log`, which stores the failed test classes from previous runs. `testQuick` combines the information from these two sources to determine the set of test classes to run.

/claim #4109

One limitation of the current implementation is after the first time `testQuick` is run (test all classes), `callGraphAnalysis` will only create the `current` folder. If we run `testQuick` again, because `callGraphAnalysis`'s inputs haven't changed, it will not run provide `previous` folder. In that case `testQuick` will not have enough information and have to run all the test classes again.

### CODE REVIEW

File: scalalib/src/mill/scalalib/JavaModule.scala [hunk 1]
```diff
@@ -1380,6 +1472,90 @@ trait JavaModule
...
+                // In Scala 3, lambdas are implemented by private instance methods,
+                // not static methods, so they fall through the crack of "isSimpleTarget".
+                // Here make the assumption that a zero-arg lambda called from a simpleTarget,
+                // should in fact be tracked. e.g. see `integration.invalidation[codesig-hello]`,
+                // where the body of the `def foo` target is a zero-arg lambda i.e. the argument
+                // of `Cacher.cachedTarget`.
+                // To be more precise I think ideally we should capture more information in the signature
+                isSimpleTarget(callSiteSig.desc) && calledSig.name.contains("$anonfun")
+              )
+            }
```
Comment (lihaoyi) [inline scalalib/src/mill/scalalib/JavaModule.scala:1521]: All the other special cases except this one are heuristics unique to `MillBuildRootModule`, and so probably should live there. We can have a `def codeSigIgnoreCall` method that gets fed into `CodeSig.compute`, with a default implementation for `JavaModule` and a more detailed override for `MillBuildRootModule`

**Comment by lihaoyi** [comment]
> One limitation of the current implementation is after the first time testQuick is run (test all classes), callGraphAnalysis will only create the current folder. If we run testQuick again, because callGraphAnalysis's inputs haven't changed, it will not run provide previous folder. In that case testQuick will not have enough information and have to run all the test classes again.

I don't understand the problem are describing here, could you go into more detail

File: scalalib/src/mill/scalalib/JavaModule.scala [hunk 2]
```diff
@@ -92,6 +95,95 @@ trait JavaModule
           case _: ClassNotFoundException => // if we can't find the classes, we certainly are not in a ScalaJSModule
         }
     }
+
+    private def quickTest(args: Seq[String]): Task[(String, Seq[TestResult])] =
+      Task(persistent = true) {
```
Comment (lihaoyi) [inline scalalib/src/mill/scalalib/JavaModule.scala:100]: `Task(persistent = true)`  is probably the wrong thing here, since `def quickTest` takes arguments. We should add a `Command(persistent = true)` flag and keep it as a command similar to the other test tasks

**Comment by lihaoyi** [comment]
> The outputs of CallGraphAnalysis now include invalidClassNames, which contains the names of test classes extracted from the spanningInvalidationTree. This list indicates the classes that are affected by code changes.

This seems like a post-processing step, could we put it in a separate method that takes the output of `spanningInvalidationTree` and parses out the parts we need?

**Comment by HollandDM** [comment]
> > One limitation of the current implementation is after the first time testQuick is run (test all classes), callGraphAnalysis will only create the current folder. If we run testQuick again, because callGraphAnalysis's inputs haven't changed, it will not run provide previous folder. In that case testQuick will not have enough information and have to run all the test classes again.
> 
> I don't understand the problem are describing here, could you go into more detail

Here is the reproduce steps:
- `rm -rf out` to clean the output folder.
- run `scalalib.callGraphAnalysis` for the first time.
- run `scalalib.callGraphAnalysis` again.

In this case, the first `callGraphAnalysis` will run normally and produce `current` folder, but  because we din't modify any source code, the second `callGraphAnalysis` will do nothing (I guess because it's input, `compile()` doesn't change). There for `previous` folder will not be produced.
The problem is that this PR will check for `previous` folder, and it will consider `testQuick` to be a first time command when it doesn't find one. So with this case especially, invoke `testQuick` consecutively will run all test cases like normal `testForked`.

=================================

> > The outputs of CallGraphAnalysis now include invalidClassNames, which contains the names of test classes extracted from the spanningInvalidationTree. This list indicates the classes that are affected by code changes.
> 
> This seems like a post-processing step, could we put it in a separate method that takes the output of `spanningInvalidationTree` and parses out the parts we need? Or could it be a different helper method that works directly on `prevTransitiveCallGraphHashes` and `transitiveCallGraphHashes0`, so we don't need to construct a `spanningInvalidationTree` at all?

I also consider post process this at first, but the output of `spanningInvalidationTree` is too easy to work with, so I went with this approach. I will update this into a separate post-processing step

**Comment by lihaoyi** [comment]
@HollandDM the problem with `callGraphAnalysis` caching should be fixed by making `testQuick` a command. That way it will run every time regardless of changes, and then can decide what to do based on whether or not tue upstream tasks changed

**Comment by lihaoyi** [comment]
> I also consider post process this at first, but the output of pre-jsonifiedspanningInvalidationTree is too easy to work with, so I went with this approach. I will update this into a separate post-processing step

Let's split it into three methods then: one shared one, one that uses the shared helper to generate the tree, and one that uses the shared helper to generate a list of affected classes

File: scalalib/src/mill/scalalib/JavaModule.scala [hunk 3]
```diff
@@ -92,6 +94,117 @@ trait JavaModule
...
+        // we cannot parse the json back to its concrete type, we only know that all of them are
+        // encoded with class name as prefix.
+        // so we can do a prefix checking to get the affected class.
+        // naive implementation won't work well as it has loop through all `testClasses` and check prefix
+        // for all of them, so we employ a prefix trie to speed up the prefix checking.
+
+        val jsonValueQueue = mutable.ArrayDeque[(String, ujson.Value)]()
+        val prefixTrie = new PrefixTrie[String]()
+        jsonValueQueue.appendAll(spanningInvalidationTreeObj.value)
+        while (jsonValueQueue.nonEmpty) {
```
Comment (HollandDM) [inline scalalib/src/mill/scalalib/JavaModule.scala:120]: > > I also consider post process this at first, but the output of pre-jsonifiedspanningInvalidationTree is too easy to work with, so I went with this approach. I will update this into a separate post-processing step
> 
> Let's split it into three methods then: one shared one, one that uses the shared helper to generate the tree, and one that uses the shared helper to generate a list of affected classes

I went with this approach first, as I think leaving `callGraphAnalysis` folder as it is right now is quite good. This pre-processing seem quick enough when I run `scalalib.test.testQuick`, so I think it is fine. WDYT?

File: core/define/src/mill/define/Task.scala [hunk 4]
```diff
@@ -121,7 +121,16 @@ object Task extends TaskBase {
...
       inline w: W[T],
       inline ctx: mill.define.Ctx
-  ): Command[T] = ${ TaskMacros.commandImpl[T]('t)('w, 'ctx, exclusive = '{ false }) }
+  ): Command[T] = ${ TaskMacros.commandImpl[T]('t)('w, 'ctx, exclusive = '{ false }, persistent = '{ false }) }
+
+
+  /**
+   * This version allow [[Command]] to be persistent
+   */
+  inline def Command[T](inline persistent: Boolean)(inline t: Result[T])(implicit
```
Comment (HollandDM) [inline core/define/src/mill/define/Task.scala:130]: add a Command with `persistent` option.

File: scalalib/src/mill/scalalib/JavaModule.scala (hunk 3 above)
Comment (lihaoyi) [reply scalalib/src/mill/scalalib/JavaModule.scala]: Looking at the code, I think splitting it into three methods is better. The various string operations necessary to do the post-processing look pretty fragile, and while it no doubt works, it probably will be harder to maintain and refactor in future compared to working with typed data structures

**Comment by lihaoyi** [comment]
I think the approach looks reasonable. Questions I have, more on the semantics and expected behavior rather than the code:

1. I assume you tested that this works for changes in the test code, but does this work for code changes in the non-test module code? Or upstream modules? The callgraph analysis we do for `MillBuildRootModule` is focused on one module, though it should in theory work for multi-module codebases too if given all the necessary classfiles

2. What kind of performance cost does this incur when used on various `.test` modules in the Mill codebase? You can run `./mill dist.installLocal && ci/patch-mill-bootstrap.sh` to get an executable you can try out locally

3. Are there any major edge cases that we should be aware of that this doesn't work well?

Eventually all this should probably make its way into a unit/integration test suite, but for now manual testing is fine just to figure out what the boundaries and limitations of this featue

**Comment by HollandDM** [comment]
For question 1, it seems that changes from upstream modules does not trigger the test. The main reason is that call graph analysis does not compute necessary data for it

**Comment by lihaoyi** [comment]
The callgraph analysis should be able to handle cross-module analysis if all the classfiles are available on disk; you should be able to aggregate the `transitiveLocalClasspath` and pass it to CodeSig and it should just work (in theory)

**Comment by HollandDM** [comment]
I've update the code to include all upstream module while running the `testQuick` command. Also clean up some shared logic.
I've also run some local test, e.g: if I update `UnitTester.scala`, then run `scalalib.test.testQuick`, all test cases that use `UnitTester` will re-run the test. This ensure cross module is working.
But for some reason, if I update both `UnitTester` and `TestRunnerTestUtils`, only changes in `TestRunnerTestUtils` are discovered. I will try to debug this

File: core/codesig/src/mill/codesig/ExternalSummary.scala [hunk 5]
```diff
@@ -47,7 +48,8 @@ object ExternalSummary {
 
     def load(cls: JCls): Unit = methodsPerCls.getOrElse(cls, load0(cls))
 
-    def load0(cls: JCls): Unit = {
+    // Some macros implementations will fail the ClassReader, we can skip them
```
Comment (lihaoyi) [inline core/codesig/src/mill/codesig/ExternalSummary.scala:51]: What kind of error do those macro implementations produce? We should try and be specific about what errors we catch here, to avoid silencing unexpected errors that may indicate real issues

File: core/codesig/src/mill/codesig/ReachabilityAnalysis.scala [hunk 6]
```diff
@@ -78,44 +73,45 @@ class CallGraphAnalysis(
...
+          transitiveCallGraphHashes0,
+          indexToNodes,
+          indexGraphEdges
+        )
+      case None => ujson.Obj()
+    }
   }
 
-  logger.mandatoryLog(spanningInvalidationTree)
+  def calculateInvalidClassName(
```
Comment (lihaoyi) [inline core/codesig/src/mill/codesig/ReachabilityAnalysis.scala:91]: Probably should call this `calculateInvalidatedClassName` to avoid confusion

File: core/codesig/src/mill/codesig/CodeSig.scala [hunk 7]
```diff
@@ -3,31 +3,63 @@ package mill.codesig
...
+    implicit val st: SymbolTable = new SymbolTable()
+
+    val callAnalysis = callGraphAnalysis(classFiles, upstreamClasspath, ignoreCall)
+
+    logger.log(callAnalysis.localSummary)
+    logger.log(callAnalysis.externalSummary)
+    logger.log(callAnalysis.resolved)
+
+    logger.mandatoryLog(callAnalysis.methodCodeHashes)
+    logger.mandatoryLog(callAnalysis.prettyCallGraph)
```
Comment (lihaoyi) [inline core/codesig/src/mill/codesig/CodeSig.scala:52]: Are these two `mandatoryLog`s necessary? It seems we only use `transitiveCallGraphHashes` and `spanningInvalidationTree`, at least as far as I can tell

File: core/codesig/src/mill/codesig/ReachabilityAnalysis.scala [hunk 8]
```diff
@@ -135,12 +131,75 @@ object CallGraphAnalysis {
...
+        indexGraphEdges
+      ),
       k => indexToNodes(k).toString
     )
   }
 
+  /**
+   * Get all class names that have their hashcode changed compared to prevTransitiveCallGraphHashes
+   */
+  def invalidClassNames(
```
Comment (lihaoyi) [inline core/codesig/src/mill/codesig/ReachabilityAnalysis.scala:170]: `invalidatedClassNames`

File: runner/src/mill/runner/MillBuildRootModule.scala [hunk 9]
```diff
@@ -138,74 +140,45 @@ abstract class MillBuildRootModule()(implicit
...
+
+    // Skip calls to `millDiscover`. `millDiscover` is bundled as part of `RootModule` for
+    // convenience, but it should really never be called by any normal Mill module/task code,
+    // and is only used by downstream code in `mill.eval`/`mill.resolve`. Thus although CodeSig's
+    // conservative analysis considers potential calls from `build_.package_$#<init>` to
+    // `millDiscover()`, we can safely ignore that possibility
+    def isMillDiscover =
+      calledSig.name == "millDiscover$lzyINIT1" ||
+        calledSig.name == "millDiscover" ||
+        callSiteOpt.exists(_.sig.name == "millDiscover")
```
Comment (lihaoyi) [inline runner/src/mill/runner/MillBuildRootModule.scala:164]: Can these be moved into `callGraphAnalysisIgnoreCalls` as well?

File: scalalib/src/mill/scalalib/JavaModule.scala [hunk 10]
```diff
@@ -1380,6 +1490,60 @@ trait JavaModule
...
+  ): Boolean = {
+    // We can ignore all calls to methods that look like Targets when traversing
+    // the call graph. We can do this because we assume `def` Targets are pure,
+    // and so any changes in their behavior will be picked up by the runtime build
+    // graph evaluator without needing to be accounted for in the post-compile
+    // bytecode callgraph analysis.
+    def isSimpleTarget(desc: mill.codesig.JvmModel.Desc) =
+      (desc.ret.pretty == classOf[mill.define.Target[?]].getName ||
+        desc.ret.pretty == classOf[mill.define.Worker[?]].getName) &&
+        desc.args.isEmpty
```
Comment (lihaoyi) [inline scalalib/src/mill/scalalib/JavaModule.scala:1507]: This should be in the `MillBuildRootModule` override method

File: scalalib/src/mill/scalalib/JavaModule.scala [hunk 11]
```diff
@@ -1380,6 +1490,60 @@ trait JavaModule
...
+          // In Scala 3, lambdas are implemented by private instance methods,
+          // not static methods, so they fall through the crack of "isSimpleTarget".
+          // Here make the assumption that a zero-arg lambda called from a simpleTarget,
+          // should in fact be tracked. e.g. see `integration.invalidation[codesig-hello]`,
+          // where the body of the `def foo` target is a zero-arg lambda i.e. the argument
+          // of `Cacher.cachedTarget`.
+          // To be more precise I think ideally we should capture more information in the signature
+          isSimpleTarget(callSiteSig.desc) && calledSig.name.contains("$anonfun")
+        )
+      }
```
Comment (lihaoyi) [inline scalalib/src/mill/scalalib/JavaModule.scala:1533]: These should be in the `MillBuildRootModule` override method I think; they look specific to Mill's own "ignore `Task`-returning methods" heuristic that won't apply to normal code

**Comment by lihaoyi** [comment]
Left some comments. Let's also add some unit tests in `scalalib.test` to exercise this logic, next to the various `TestRunnerTests` that exercise the rest of the testing logic:

1. single-module Java project
2. single-module Scala project
3. multi-module Java project
4. multi-module Scala project

Since we don't need to make any changes to the `build.mill` to exercise this logic, unit tests are fine and we don't need an integration test.

File: scalalib/src/mill/scalalib/JavaModule.scala (hunk 11 above)
Comment (HollandDM) [reply scalalib/src/mill/scalalib/JavaModule.scala:1533]: If we allow analysis to go through `Task`-liked classes/objects, then by design it will spread into all `task`s. In `mill` project there is tests that create a custom module of `Test Module` (`TestModuleTestUtils`, etc...). So in this case it will trigger every `task`s.
Should we keep it like this, or allow `testQuick` to trigger `task`s like that?

File: scalalib/src/mill/scalalib/PrefixTrie.scala [hunk 12]
```diff
@@ -0,0 +1,66 @@
+package mill.scalalib
+
+import scala.collection.mutable
+
+private[scalalib] final class PrefixTrie[A](using CanEqual[A, A]) {
```
[the remark this answers is not in the document]
Comment (HollandDM) [reply scalalib/src/mill/scalalib/PrefixTrie.scala:5]: True, I must have missed it

File: core/codesig/src/mill/codesig/CodeSig.scala (hunk 7 above)
Comment (HollandDM) [reply core/codesig/src/mill/codesig/CodeSig.scala:52]: I have no ideal either, I put them here to mimic the original code, which log everything like this

File: scalalib/src/mill/scalalib/JavaModule.scala (hunk 11 above)
Comment (lihaoyi) [reply scalalib/src/mill/scalalib/JavaModule.scala:1533]: The issue is that this problem exists for other interfaces as well. For Mill and Mill plugins the heuristic is to ignore `Task`, but for other projects with other base abstractions the heuristic would need to be something different to be useful. Let's think if we can come up with a more general heuristic that isn't hardcoded to depend on Mill-specific types like `Task`, using the data we already have

What if rather than using a method-based callgraph, we did the analysis on a more coarse grained class-based dependency graph? This is something we can't really do for invalidating tasks since tasks are individual methods, but since test selection is at a granularity of classes that might work. Presumably if `ClassA` never directly or transitively references `ClassB`, and `ClassA` doesn't take any constructor parameters (since it's a test suite class and those usually have zero parameters) then changes to `ClassB` should never affect `ClassA`? This feels like an approach that could either replace the existing method-level approach, or be stacked on top as an additional heuristic (e.g. only invalidate a test or task if both the method-level codesig and the class-level codesig both change)

Comment (HollandDM) [reply scalalib/src/mill/scalalib/JavaModule.scala:1533]: > The issue is that this problem exists for other interfaces as well. For Mill and Mill plugins the heuristic is to ignore Task, but for other projects with other base abstractions the heuristic would need to be something different to be useful. Let's think if we can come up with a more general heuristic that isn't hardcoded to depend on Mill-specific types like Task, using the data we already have

This is true, we do need to stop using Mill-specific heuristic

**Comment by HollandDM** [comment]
@lihaoyi I've update the code to address the comments and answer some questions.

For using class based graph instead of method call, I've tried using the class dep graph produced from zinc as reference, while it works, I still need to use the `ReachabilityAnalysis` to check the hash code (not sure if zinc can do it itself). And after some thoughts, while leaving method call to pass the `Task` boundary result in a bigger call graph, in the end we only check if the method's class name  contains the module test classes discovered from `discoverTestClasses`, so the test classes that need to be run will never more than the normal `testForked`. So I guess leave it as it is for now is also good enough.

If the code is good to go, I'll proceed with writing test cases

**Comment by lihaoyi** [comment]
Thanks @HollandDM ! will take a look

**Comment by lihaoyi** [comment]
I think the code looks reasonable, at least as I would expect.

However I would like to push a bit more on the design of the feature. How does Zinc implement `testQuick`? Presumably they must have made similar tradeoffs as we are considering here. We should find out what approach they chose and why they chose it, so we can make a better decision rather than just arbitrarily picking a design on our own.

**Comment by HollandDM** [comment]
Not sure what zinc's `testQuick` is, I don't think zinc has that, but `sbt` does have `testQuick` command.
If going zinc way, I think we can use the output analysis of previous `compile` and current `compile`, compare there `stamps` to extract changes classes (`product`s), then use the current analysis again to get all relations, and do something similar to spanning forest on the graph.
Comparing zinc stamp meaning we'll compare mostly the hash of the whole file content.

**Comment by HollandDM** [comment]
@lihaoyi this is also updated to include the similar integration tests

File: integration/feature/test-quick/src/TestQuickTests.scala [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        val exists = firstRunOutLines.exists(_.contains(expectedLines))
+        assert(exists)
+      }
+
+      // Second run, nothing should run because we're not changing anything
+      val secondRun = eval("app.test.testQuick")
+      assert(secondRun.out.isEmpty)
+
+      // Third run, MyNumber.scala changed, so MyNumberDefaultValueTests should run
+      modifyFile(
```
Comment (HollandDM) [inline integration/feature/test-quick/src/TestQuickTests.scala:30]: This is one win of this approach, because we do a more fine grain graph invalidation, only relevant test class will need to be rerun

File: integration/feature/test-quick/src/TestQuickTests.scala [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        val exists = firstRunOutLines.exists(_.contains(expectedLines))
+        assert(exists)
+      }
+
+      // Second run, nothing should run because we're not changing anything
+      val secondRun = eval("app.test.testQuick")
+      assert(secondRun.out.isEmpty)
+
+      // Third run, Combinator.scala changed syntactically, should not run
+      modifyFile(
```
Comment (HollandDM) [inline integration/feature/test-quick/src/TestQuickTests.scala:95]: another win for this approach is that it work on JVM bytecode, so some syntax changes do not invalidate node in graph, although source hashcode can be changed

**Comment by HollandDM** [comment]
I run this with time command for a few time, here is the result:

./mill-assembly.jar --no-server scalalib.test.testForked: 147,53s; 148.53s; 151.19s (avg: 149.08s)
./mill-assembly.jar --no-server scalalib.test.testQuick: 154.19s; 150.36s; 169.86s (avg: 158.14s)

So in average, the cost to calculate the re-run test is about ~9 extra seconds. Compare to the overall runtime of the command, this is about ~5.8% slow down.
The testQuick was triggered by modifying UnitTester.scala file, which will invalidate all test classes

**Comment by HollandDM** [comment]
@lihaoyi the example test and adoc are pushed, please have a look.
For some reason, I cannot make Java example to run correctly, despite Scala and Kotlin does. I will look into further when rebasing.
But I do want to ask for your experience on Java `JupiterFramework` framework. In the example, the test run correctly until the line `> mill -j 1 foo.test.testQuick # Fourth run re-runs failing FooTest2 even without code changes`. Some debugging yields that `FooTest2.java` is not even run, hence it cannot throw error.

**Comment by lihaoyi** [comment]
@HollandDM I know that Junit5/Jupiter has some special casing in most build tools, including Mill, not sure if that's what you're seeing, but AFAIK it should take place entirely before the test filtering logic done by `testQuick` so I don't see how it could affect your test:

- https://github.com/com-lihaoyi/mill/blob/d2a31808da38682dba74f118209bc842f30b0382/libs/scalalib/src/mill/scalalib/TestModule.scala#L374

Is `FooTest2` not being run the second time, or is it not being run at all even the first time? Is it run if you use normal `testForked`?

File: example/javalib/testing/7-test-quick/build.mill [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+  }
+  // Ultilities for replacing text in files
+  def replaceBar(args: String*) = Task.Command {
+    val relativePath = os.RelPath("../../../foo/src/Bar.java")
+    val filePath = Task.dest() / relativePath
+    os.write.over(filePath, os.read(filePath).replace(
+      """return String.format("Hi, %s!", name);""",
+      """return String.format("Ciao, %s!", name);"""
+    ))
+  }
```
Comment (lihaoyi) [inline example/javalib/testing/7-test-quick/build.mill:22]: Let's move these into the `/** Usage` blocks using `sed -i.bak` rather than having it as part of the Scala code; can grep for that string to see examples of that in the codebase

**Comment by HollandDM** [comment]
> Is `FooTest2` not being run the second time, or is it not being run at all even the first time? Is it run if you use normal `testForked`?
It does run the first time, and stop running the second time if there is no update
